ggml-cuda.cu

```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
+        CUDA_CHECK(cudaGetLastError());
+    }
+
+    (void) dst;
+    (void) src0_ddq_i;
+}
+
+inline void ggml_cuda_op_dequantize_mul_mat_vec(
+    const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst, char * src0_ddq_i,
+    float * src0_ddf_i, float * src1_ddf_i, float * dst_ddf_i, uint64_t i0_low, uint64_t i0_high, int i1,
+    cudaStream_t & cudaStream_main){
+
+    GGML_ASSERT(src0_ddq_i != nullptr);
+    GGML_ASSERT(src1_ddf_i != nullptr);
+    GGML_ASSERT(dst_ddf_i != nullptr);
+
+    const int64_t ne00 = src0->ne[0];
+    const int64_t nrows = i0_high - i0_low;
+
+    switch (src0->type) {
+        case GGML_TYPE_Q4_0:
+            dequantize_mul_mat_vec_q4_0_cuda(src0_ddq_i, src1_ddf_i, dst_ddf_i, ne00, nrows, cudaStream_main);
+            break;
+        case GGML_TYPE_Q4_1:
+            dequantize_mul_mat_vec_q4_1_cuda(src0_ddq_i, src1_ddf_i, dst_ddf_i, ne00, nrows, cudaStream_main);
+            break;
+        case GGML_TYPE_Q5_0:
+            dequantize_mul_mat_vec_q5_0_cuda(src0_ddq_i, src1_ddf_i, dst_ddf_i, ne00, nrows, cudaStream_main);
+            break;
+        case GGML_TYPE_Q5_1:
+            dequantize_mul_mat_vec_q5_1_cuda(src0_ddq_i, src1_ddf_i, dst_ddf_i, ne00, nrows, cudaStream_main);
+            break;
+        case GGML_TYPE_Q8_0:
+            dequantize_mul_mat_vec_q8_0_cuda(src0_ddq_i, src1_ddf_i, dst_ddf_i, ne00, nrows, cudaStream_main);
+            break;
+        case GGML_TYPE_F16:
+            convert_mul_mat_vec_f16_cuda(src0_ddq_i, src1_ddf_i, dst_ddf_i, ne00, nrows, cudaStream_main);
+            break;
+        default:
+            GGML_ASSERT(false);
+            break;
+    }
+    CUDA_CHECK(cudaGetLastError());
+
+    (void) src1;
+    (void) dst;
+    (void) src0_ddf_i;
+    (void) i1;
+}
+
+inline void ggml_cuda_op_mul_mat_cublas(
+    const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst, char * src0_ddq_i,
+    float * src0_ddf_i, float * src1_ddf_i, float * dst_ddf_i, uint64_t i0_low, uint64_t i0_high, int i1,
+    cudaStream_t & cudaStream_main){
+
+    GGML_ASSERT(src0_ddf_i != nullptr);
+    GGML_ASSERT(src1_ddf_i != nullptr);
+    GGML_ASSERT(dst_ddf_i != nullptr);
+
+    const float alpha = 1.0f;
+    const float beta = 0.0f;
+
+    const uint64_t ne00 = src0->ne[0];
+
+    const uint64_t ne10 = src1->ne[0];
+    const uint64_t ne11 = src1->ne[1];
+
+    const uint64_t i0_diff = i0_high - i0_low;
+
+    CUBLAS_CHECK(cublasSetStream(g_cublasH, cudaStream_main));
+    CUBLAS_CHECK(
+        cublasSgemm(g_cublasH, CUBLAS_OP_T, CUBLAS_OP_N,
+                i0_diff, ne11, ne10,
+                &alpha, src0_ddf_i, ne00,
+                        src1_ddf_i, ne10,
+                &beta,  dst_ddf_i,  i0_diff));
+
+    (void) dst;
+    (void) src0_ddq_i;
+    (void) i1;
+}
+
+template<enum ggml_cuda_op_type op_type, ggml_cuda_op_t op>
+static void ggml_cuda_op(const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst) {
```

Comment:
I am not sure that there is a good reason to make this a template, the performance hit of doing these checks at runtime should be negligible. I would also consider replacing `ggml_cuda_op_type` with explicit checks for the type (from what I understand, the way this is used you could simply use `tensor->type`).

Reply:
It would also remove the need to use magic bit checks like this:
https://github.com/ggerganov/llama.cpp/blob/bee33f71759299591f85b7bc0d5622f1e64d724e/ggml-cuda.cu#L782

Reply:
A template is a little more convenient for debugging since the NVIDIA compute sanitizer will then give you a better stacktrace but I would be fine with a regular function.

Reply:
With nsight systems it is possible to use NVTX to add markers, it may also work with the compute sanitizer, not sure.

Reply:
How would changing `ggml_cuda_op` from a template help with the check to determine whether `src0` requires f32? The check uses `GGML_CUDA_OP_TYPE` which would still have to be passed as a parameter, right?

https://github.com/JohannesGaessler/llama.cpp/blob/f22982f84b86fd28f30f61d0c696d7d5f6efa507/ggml-cuda.cu#L36-L45

Reply:
Right, I wasn't sure if the `op_type` is needed at all. Still, I would suggest replacing `ggml_cuda_op_type` with explicit types to be more clear. Rather than an enum, this is really a struct with 3 booleans. More generally, you could replace it with a struct such as:
```C++
struct ggml_cuda_op_type {
    ggml_type src0_type, src1_type, dst_type;
}
```
Or just pass `src0_type` etc as individual parameters to the function, it doesn't matter. The goal here is to improve code clarity and remove possible sources of errors in the future.

Reply:
For now I've just made it a boolean since only `src0` can be in need of quantization; it shouldn't be too hard to change if it becomes necessary.